### ISSUE TITLE
Potential bug of model output dim in reward trainer example

### ISSUE THREAD

**Issue by liutianlin0121** [lvwerra/trl#558]
Hi!

I've been exploring the `examples/scripts/reward_trainer.py` and believe I've found a potential bug that I'd like to address.

## Potential bug.
To my understanding, the reward model should produce a scalar output. However, when we load the pre-trained model in the [example script](https://github.com/lvwerra/trl/blob/main/examples/scripts/reward_trainer.py#L75-L80), the model will be a binary classifier by default, with a 2-dim logit output:

```python
import torch
inputs = tokenizer("Hello, my dog is cute", return_tensors="pt")
model = model.cpu()
with torch.no_grad():
    out = model(**inputs)
print(out.logits.shape) # torch.Size([1, 2])
```

During training, this issue was hidden by the [the loss computation part](https://github.com/lvwerra/trl/blob/main/trl/trainer/reward_trainer.py#L169-L176) of the reward trainer:

```python
    rewards_chosen = model(
        input_ids=inputs["input_ids_chosen"],
        attention_mask=inputs["attention_mask_chosen"],
    )[0]

    rewards_rejected = model(
        input_ids=inputs["input_ids_rejected"],
        attention_mask=inputs["attention_mask_rejected"],
    )[0]

    loss = -torch.nn.functional.logsigmoid(rewards_chosen - rewards_rejected).mean()
```
There, the `rewards_rejected` and `rewards_chosen` are dim [sample, 2] matrices, which I think should be dim [sample, 1] instead. The `.mean()` operator hides the issue as it returns a scalar value.

Here is a [colab](https://colab.research.google.com/drive/1szzGC8H1DQ7329_idVnnOnPIPE6D7J4j?usp=sharing) that shows the dimensionality problem.

## Proposed fix

To resolve this, in the model training example, we can use:

```python
model = AutoModelForSequenceClassification.from_pretrained(
    script_args.model_name,
    quantization_config=quantization_config,
    device_map=device_map,
    trust_remote_code=script_args.trust_remote_code,
    num_labels=1 # <---- make sure that the model output is a scalar
)
```
I'll submit a pull request if it is OK. 

In addition, while exploring `examples/scripts/reward_trainer.py`, I noticed a few other places that can be improved. For example, there is currently no validation steps that show the genearlization performance of the reward model; the `log_with` option in `ScriptArguments` is currently not being used anywhere and therefore has no effect. I'm happy to include the validation and fix the logging glitch in my pull request as well.

Tianlin

**Comment by younesbelkada** [comment]
Hi @liutianlin0121 
Thanks for the issue, I went through the shared materials and details. I also have checked with @edbeeching and we think that you are right, the proposed fix seems to be the correct one. Would you ne happy opening a PR for the fix?

**Comment by liutianlin0121** [comment]
Thanks! Yes I'll submit a PR soon :-)
